CI: Check for dependency updates automatically

hoffie opened the pull request:
<!-- Thank you for working on Jamulus and opening a Pull Request! Please fill in the following to make the review process straightforward -->

**Short description of changes**
This PR adds a workflow which automatically checks many of our pinned third-party dependencies which cannot be handled easily by dependabot. It currently covers:
- [x] aqt, Example PR: https://github.com/hoffie/jamulus/pull/85 ([Logs](https://github.com/hoffie/jamulus/runs/7805028689?check_suite_focus=true#step:3:63))
- [x] Qt6, Example PR: https://github.com/hoffie/jamulus/pull/86 ([Logs](https://github.com/hoffie/jamulus/runs/7805028808?check_suite_focus=true#step:3:65))
- [x] choco-jack, Example PR: https://github.com/hoffie/jamulus/pull/83 ([Logs](https://github.com/hoffie/jamulus/runs/7805028917?check_suite_focus=true#step:3:65))
- [x] choco-jom, Example PR: https://github.com/hoffie/jamulus/pull/81 ([Logs](https://github.com/hoffie/jamulus/runs/7805029019?check_suite_focus=true#step:3:65))
- [x] NSIS, Example PR: https://github.com/hoffie/jamulus/pull/82 ([Logs](https://github.com/hoffie/jamulus/runs/7805029124?check_suite_focus=true#step:3:65))
- [x] ASIO-SDK, Example PR: https://github.com/hoffie/jamulus/pull/84 ([Logs](https://github.com/hoffie/jamulus/runs/7805029239?check_suite_focus=true#step:3:65))

(The logs are from runs on my repo which had artificially been made outdated regarding all supported components before)

If an update is found, a PR is submitted automatically. The workflow automatically runs once per week (Saturdays).
Upon merge or close of the PR, the branch is deletec automatically.
PRs are updated automatically if `master` changes and there is a risk of conflicts.
The workflow has been made re-usable and uses a matrix to check the different targets, so it should be somewhat easy to extend (once one has come up with a proper regexp...).

The workflow does not handle most of the Android pinnings (haven't looked into it so far).
It does not handle submodules yet (dependabot supports them, but not based on releases/tags, it seems).
It does not handle pinned Github Actions, as those should be compatible with

CHANGELOG: SKIP

**Context: Fixes an issue?**
Related: #2346

<!-- If this fixes an issue, please write Fixes: <issue number here>; if not, please give your PR a context. -->

**Does this change need documentation? What needs to be documented and how?**

It mostly needs inline documentation to avoid accidential breakage during future refactorings. This documentation in the form of comments is included in this PR.

**Status of this Pull Request**

Ready for review.

**What is missing until this pull request can be merged?**
- [x] Reviews
- [x] Merge of #2665 to avoid unnecessary duplicate PRs
- [x] We need to generate an SSH deploy key
- [x] Add the public key to jamulussoftware/jamulus with write access
- [x] Add the private key as repository secret to jamulussoftware/jamulus
(The deploy key is necessary as the workflow is not allowed to modify .github/workflows/ by itself, even in a PR; it's also necessary to ensure proper Checks/Autobuilds on those bot-submitted PRs)


<!-- Does it still need more testing; ... -->

## Checklist

<!-- Please tick the check boxes when done by replacing the space by an x, e.g. [x]. -->

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) <!-- You can also check if your code passes clang-format -->
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. <!-- GitHub doesn't run these checks for new contributors automatically. -->
-  [x] I've filled all the content above

ann0see commented on .github/workflows/bump-dependencies.yml:
```diff
@@ -0,0 +1,174 @@
+  # For PR creation:
+  workflow_dispatch:
+  schedule:
+    - cron: '13 8 * * 6'  # once a week on Saturdays at 08:13
```
Any reason for that time? Usually maintenance takes place at night, I think? (It doesn't matter much though...).

hoffie replied:
It's rather arbitrary. I'd avoid the standard minutes 0/15/30/45 as those are usually loaded (at least in my experience from managing larger fleets of servers). Haven't thought about the exact hour a lot, but my thinking was that main usage is probably dominated by other timezones anyway.

ann0see commented on .github/workflows/bump-dependencies.yml:
```diff
@@ -0,0 +1,174 @@
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          ssh-key: ${{ secrets.BUMP_DEPENDENCIES_SSH_DEPLOY_KEY || 'fail-due-to-missing-ssh-key-as-secret' }}
```
@ignotus666 how did you handle pushing to the website app without SSH key? I'd prefer that. Maybe it doesn't work well with branch protections, but you don't push to a protected branch, I think.

ann0see replied:
Ah ok. @hoffie I just read your hint on that. Maybe we can still find a way around that (we might want to use a JSON based Matrix anyway soon to allow a dynamically created matrix).

hoffie replied:
> Maybe we can still find a way around that

The officially recommended workaround is using a bot user. I postpone that topic for Jamulus somehow as it would require creating a user, managing/rotating its password, storing its password somewhere for Team access (Github secrets?), finding a way to handle 2FA for the whole team. Also, such users (and their Personal Access Tokens) don't seem to have fine-grained permissions. You can choose the permissions, but not the subjects (i.e. repos), for example.

Yet another option is creating repo-specific Github apps, which sounds unnecessarily complicated to me.

Therefore, this seemed like the most-straightforward solution to allow pushing to .github/workflows/ and to trigger regular PR workflows (Autobuild).

pljones commented on .github/workflows/bump-dependencies.yml:
```diff
@@ -0,0 +1,179 @@
+
+jobs:
+  create-prs:
+    name: Update ${{ matrix.components.name }} if necessary
```
These need to be clearly from an automated source.  Flagging that in the title would be good.

pljones commented on .github/workflows/bump-dependencies.yml:
```diff
@@ -0,0 +1,179 @@
+          # Therefore, use perl instead:
+          perl -pe 's/${{ matrix.components.local_version_regex }}/${1}'"${upstream_version}"'${3}/gi' -i "${files[@]}"
+          git add .
+          title="Build: Update ${{ matrix.components.name }} to ${upstream_version}"
```
I'd like the title to make it clear this is automation.  The body could have SKIP, really.

hoffie replied:
I've updated the code to append `(Automated PR)` to the PR title.
Did you want to have such a comment in the commit as well? I'd rather not have it there. If the commit makes it into `master`, it will have been reviewed by a human. As such, I don't think it's worth denoting it's automated source (besides the author info, which covers it anyway).

ann0see replied:
I think we Prefix the po file update with AUTO: on the website (not sure). We should be consistent.

hoffie replied:
Yes, but the difference is that the .po files are [auto-committed](https://github.com/jamulussoftware/jamuluswebsite/blob/release/.github/workflows/main.yml#L41-L46) to a non-feature branch, aren't they?
In contrast, we are talking about auto-PRs here (with no intention to do that unattended).

pljones replied:
I'd still rather have something other than `Build` as the commit comment prefix that makes it simple to search for these changes.

hoffie replied:
Something like [`git log --author=actions@github.com`](https://github.com/hoffie/jamulus/commit/98d2a7ea8cfdae2365b4bd4b3e11602e00e3ac75)? ;)

```
$ git log --author=actions@github.com
commit 98d2a7ea8cfdae2365b4bd4b3e11602e00e3ac75 (HEAD, hoffie/ci/bump-dependencies/aqt)
Author: github-actions[bot] <actions@github.com>
Date:   Sat Aug 13 11:18:16 2022 +0000

    Build: Update aqt to 2.2.1
```

Or do you still think we should add the AUTO: prefix here? I still think we shouldn't, but in the end it's better to add it instead of not merging this PR. ;)

ann0see replied:
I don't have a strong opinion here. Yes, the author should make the fact that it's automated clear.

pljones replied:
> Something like [git log --author=actions@github.com](https://github.com/hoffie/jamulus/commit/98d2a7ea8cfdae2365b4bd4b3e11602e00e3ac75)? ;)

Hm... I guess that would look something like
```
98d2a7ea 2022-08-13 Build: Update aqt to 2.2.1 [github-actions[bot]]
340ea336 2022-08-09 (HEAD -> master, origin/master, origin/HEAD) Set version to 3.9.0dev [ann0see]
983ae3d3 2022-08-09 (tag: r3_9_0nightly1) r3.9.0Nightly1 [ann0see]
820c4aa6 2022-08-09 Merge pull request #2774 from softins/fix-new-empty-chaninfo [ann0see]
```
for my git log.  So yes, I can see it was a bot and what was done.  That'll do.

ann0see reviewed (approved):
Looks good. Thanks!

hoffie commented:
Generated a new ed25519 keypair, added the private key as repository secret, added the public key as deploy key with write permissions. I've destroyed my local copy of the key. It is not needed. I don't think we can easily access the private key again, but we shouldn't have to. If something changes, we'll just generate a new one and change the public key.

hoffie commented:
... and it did work: #2787
Autobuild CI started properly as well.

ann0see commented:
> GraphQL: Resource not accessible by integration (repository.pullRequest.projectCards.nodes.0)

There's an error on master.

hoffie commented:
Yep, I've seen it and I'm currently looking into it. Might be permission-related, but I'm not yet sure why.

hoffie commented:
I guess `gh` automatically tries to resolve linked resources, in this case: projects. I did not grant read access for that yet. Will try to reproduce that on my repo and will submit a PR if confirmed.

hoffie commented:
Meh, I can't reproduce it on my repo. I think the issue is that we use organization-scoped project boards and that the {{GITHUB_TOKEN}} is repo-scoped (there's no way around that besides using PATs). I'll try to see if I can find a way to do the necessary changes without requiring that permission, but I will have to run the tests in jamulussoftware/jamulus (not on master, though).
